src/store/modules/segmentationModule.js

```diff
@@ -0,0 +1,1411 @@
+ *                             `Labelmap3D` on the element.
+ *
+ * @param  {HTMLElement} element  The cornerstone enabled element.
+ * @param  {string} increaseOrDecrease = Whether to increase/decrease the activeLabelmapIndex.
+ * @returns {null}
+ */
+function _changeActiveSegmentIndex(element, increaseOrDecrease = 'increase') {
+  const stackState = getToolState(element, 'stack');
+  const stackData = stackState.data[0];
+  const firstImageId = stackData.imageIds[0];
+
+  const brushStackState = state.series[firstImageId];
+
+  if (!brushStackState) {
+    return;
+  }
+
+  const activeLabelmapIndex = brushStackState.activeLabelmapIndex;
+  const labelmap3D = brushStackState.labelmaps3D[activeLabelmapIndex];
+
+  switch (increaseOrDecrease) {
+    case 'increase':
+      labelmap3D.activeSegmentIndex++;
+
+      if (labelmap3D.activeSegmentIndex > state.segmentsPerLabelmap) {
+        labelmap3D.activeSegmentIndex = 1;
+      }
+      break;
+    case 'decrease':
+      labelmap3D.activeSegmentIndex--;
+
+      if (labelmap3D.activeSegmentIndex <= 0) {
+        labelmap3D.activeSegmentIndex = state.segmentsPerLabelmap;
+      }
+      break;
+  }
+}
+
+/**
+ * _addLabelmap3D - Adds a `Labelmap3D` object to the `BrushStackState` object.
+ *
+ * @param  {BrushStackState} brushStackState The labelmap state for a particular stack.
+ * @param  {number} labelmapIndex   The labelmapIndex to set.
+ * @param  {number} size            The size of the ArrayBuffer in bytes/ 2.
+ * @returns {null}
+ */
+function _addLabelmap3D(brushStackState, labelmapIndex, size) {
+  logger.warn('hmm..');
+  logger.warn(size);
+
+  // Buffer size is multiplied by 2 as we are using 2 bytes/voxel for 65536 segments.
+  brushStackState.labelmaps3D[labelmapIndex] = {
+    buffer: new ArrayBuffer(size * 2),
+    labelmaps2D: [],
+    metadata: [],
+    activeSegmentIndex: 1,
+    segmentsVisible: [],
+  };
+}
+
+/**
+ * _addLabelmap2DView - Adds a `Labelmap2D` view of one frame of a `Labelmap3D`.
+ *
+ * @param  {BrushStackState} brushStackState     The `BrushStackState` for a particular `Series`.
+ * @param  {number} labelmapIndex       The labelmap index.
+ * @param  {number} imageIdIndex        The stack position of the image.
+ * @param  {number} rows                The number of rows in the image.
+ * @param  {number} columns             The number of columns in the image.
+ * @returns {null}
+ */
+function _addLabelmap2DView(
+  brushStackState,
+  labelmapIndex,
+  imageIdIndex,
+  rows,
+  columns
+) {
+  const sliceLength = rows * columns;
+  const byteOffset = sliceLength * 2 * imageIdIndex; // 2 bytes/pixel
+
+  const pixelData = new Uint16Array(
+    brushStackState.labelmaps3D[labelmapIndex].buffer,
+    byteOffset,
+    sliceLength
+  );
+
+  brushStackState.labelmaps3D[labelmapIndex].labelmaps2D[imageIdIndex] = {
+    pixelData,
+    segmentsOnLabelmap: [],
+    invalidated: true,
+  };
+}
```

Comment:
This file is huuuuge

Reply:
Started to skim toward the end, but my general impression is:

- Simplify function names (drop get/set prefixes)
- DRY up the duplicate logic around checks and fetching state
- Where it makes sense, reduce `null` and `undefined` checks using smarter data initialization
    - Maybe with calls to "addLabelmap" that include logic like that
- Module description that examples data shape / usage? (existing typedefs were helpful, but why/when/where I would use certain methods, from an end user's perspective, were not immediately obvious)

Reply:
Happy for the feedback, yeah it is massive currently, we should discuss this today if you have time:

> Simplify function names (drop get/set prefixes)

Some have both a setter and getter, the exported function names are tidied up and on the correct `getters` or `setters` object, but for a lot of these I'm not sure how you'd remove them without making it unclear.

> DRY up the duplicate logic around checks and fetching state

Yeah there is certainly some work that can be done here, I'll get on it after the other big refactors remaining.

> Where it makes sense, reduce null and undefined checks using smarter data initialisation
The main idea was that we only want to define a `Labelmap2D` view if they actually contain data, and remove them if they no longer do, for two reasons:
1) If there is no data on the current imageId, we can jump out early on the render to canvas.
2) when you wish to export you only have to pull data from the `Labelmap2D` objects that are defined, rather than scouring the whole volume.

We could define them all upon initialisation of the `Labelmap3D`, and then add an 'empty' (or `hasSegments`) flag for each `Labelmap2D`, but perhaps this change would be more book keeping, rather than less? Not sure 🤔 .

> Module description that examples data shape / usage? (existing typedefs were helpful, but why/when/where I would use certain methods, from an end user's perspective, were not immediately obvious)

I was going to get around to this once I write the docs for this module, do you think we need something other than documentation here?